c10/cuda/CUDACachingAllocator.h

```diff
@@ -117,38 +117,218 @@ struct SegmentInfo {
+C10_CUDA_API DeviceStats getDeviceStats(int device); \
+C10_CUDA_API void resetAccumulatedStats(int device); \
+C10_CUDA_API void resetPeakStats(int device); \
+C10_CUDA_API std::vector<SegmentInfo> snapshot(); \
+C10_CUDA_API void notifyCaptureBegin(int device, CaptureId_t graph_id, MempoolId_t mempool_id); \
+C10_CUDA_API void notifyCaptureAboutToEnd(int device, CaptureId_t graph_id); \
+C10_CUDA_API void notifyCaptureEnded(int device, CaptureId_t graph_id); \
+C10_CUDA_API void notifyCaptureDestroy(int device, MempoolId_t mempool_id); \
+C10_CUDA_API std::mutex* getFreeMutex(); \
+C10_CUDA_API std::shared_ptr<void> getIpcDevPtr(std::string handle);
+
+// Not meant to be called directly by clients.
+// Maybe make "CUDACachingAllocator" a class or struct, and make these private members?
+namespace Native {
+CUDA_ALLOCATOR_BACKEND_INTERFACE
+}
+
+// Not meant to be called directly by clients.
+namespace CudaMallocAsync {
+CUDA_ALLOCATOR_BACKEND_INTERFACE
+}
+
+// The following functions ARE meant to be called directly by clients.
+// They'll choose the appropriate backend based on the runtime value of
+// the PYTORCH_CUDA_ALLOC_CONF environment variable
+// (cf parseArgs in CUDACachingAllocator.cpp)
+
+inline void* raw_alloc(size_t nbytes) {
+  // Lean on the out-of-line call here as the surface at which we choose the allocator backend
+  // The selecting conditional could easily be made smarter, ie, it could be a lambda like
+  // static auto f = [] {
+  //     if (std::strcmp(allocatorBackend(), "native") == 0) {
+  //       return CudaMallocAsync::raw_alloc;
+  //     } else if (backend == "cudaMallocAsync") {
+  //       return CudaMallocAsync::raw_alloc;
+  //     } else {
+  //       Assume "backend" is the name of a user-supplied .so.
+  //       Use dlopen+dlsym to fish the raw_alloc symbol from it.
+  //       return raw_alloc_symbol_from_user_lib;
+  //     }
+  //   }();
+  static auto f = (std::strcmp(allocatorBackend(), "native") == 0) ?
+    Native::raw_alloc : CudaMallocAsync::raw_alloc;
+  return f(nbytes);
+  // The downside of ^ is that it's not inlineable even with LTO.
+  // If LTO is enabled and we think it's capable of inlining any allocator backend's calls at the
+  // point of use, we could just use if statements and hope for good branch prediction.
+  // static bool useNative = (allocatorBackend() == "native");
+  // if (useCudaMallocAsync) {
+  //   return CudaMallocAsync(raw_alloc);
+  // } else {
+  //   return Native::raw_alloc(nbytes);
+  // }
+}
+
+inline void* raw_alloc_with_stream(size_t nbytes, cudaStream_t stream) {
+  static auto f = (std::strcmp(allocatorBackend(), "native") == 0) ?
+    Native::raw_alloc_with_stream : CudaMallocAsync::raw_alloc_with_stream;
+  return f(nbytes, stream);
+}
+
+inline void raw_delete(void* ptr) {
+  static auto f = (std::strcmp(allocatorBackend(), "native") == 0) ?
+    Native::raw_delete : CudaMallocAsync::raw_delete;
+  return f(ptr);
```

Comment:
@swolchok Do you know if there's a better idiom for doing this in C++? Basically, mcarilli wants to have two implementations of the same interface, but he wants to select between the two via a predictable branch rather than a normal dynamic dispatch (which is how you would ordinarily do it).

Reply:
> predictable branch rather than a normal dynamic dispatch 

branch predictors in modern processors should be able to handle indirect jumps (which should be what implements virtual function calls). do you have measurements saying they can't?

Reply:
citation: https://www.agner.org/optimize/microarchitecture.pdf warns that it doesn't work so great in processors older than the Pentium M (vintage 2003) and K10 (vintage 2007) in section 3.15 .

Reply:
I don't. So we could probably just do a normal virtual interface here then!

Reply:
the one caveat is that virtual function calls will prohibit inlining, whereas a branch may not. If this level of performance/size matters to you then you'll want the branch.